### PR TITLE
Convert IATS Settings to be standard practice

### PR DESCRIPTION
**Pull request by seamuslee001**
This is a resubmit of #437 and aims to fix the issue with the iats_days setting not being an array consistently as well @KarinG are you able to run this on the Webform CiviCRM tests please.

ping @adixon

### CODE REVIEW

**Comment by KarinG** [comment]
Dave ran a one off last week that showed #437 + your ad-hoc PR pass our WFC matrix. @seamuslee001 - does JMA have this #437 + your ad-hoc PR in production? It would be good to confirm all is well with some live testing.

**Comment by KarinG** [comment]
Re-running with:

image

**Comment by KarinG** [comment]
@adixon -> this PR works with all our WFC iATS testing, including recurring:

image

**Comment by seamuslee001** [comment]
@KarinG @adixon any issues with merging this now do you think?

File: iats.php [hunk 1]
```diff
@@ -287,6 +287,19 @@ function iats_civicrm_buildForm_CRM_Contribute_Form_Contribution_Main(&$form) {
   return iats_civicrm_buildForm_CRM_Financial_Form_Payment($form);
 }
 
+/**
+ * Implements hook_civicrm_buildForm for the CRM_Admin_Form_Generic form
+ *
+ * @var CRM_Admin_Form_Generic $form
+ */
+function iats_civicrm_buildForm_CRM_Admin_Form_Generic($form) {
+  if ($form->getSettingPageFIlter() === 'iats') {
```
[the remark this answers is not in the document]
Comment (adixon) [reply iats.php:296]: Note to self - this is replacing the rules part of the old form/file: CRM/Iats/Form/Settings.php

[the remark this answers is not in the document]
Comment (adixon) [reply iats.php:296]: Any reason you wouldn't want to implement this function's functionality via a validate_callback in the iats.settings.php array? As it is, you're not really freeing us up from quickform, but if it's for simplicity as a first step, then I understand ...

File: xml/Menu/iats.xml [hunk 2]
```diff
@@ -12,8 +12,8 @@
     <access_arguments>make online contributions</access_arguments>
   </item>
   <item>
-    <path>civicrm/admin/contribute/iatssettings</path>
-    <page_callback>CRM_Iats_Form_Settings</page_callback>
+    <path>civicrm/admin/setting/iats</path>
+    <page_callback>CRM_Admin_Form_Generic</page_callback>
```
Comment (adixon) [inline xml/Menu/iats.xml:16]: As per earlier note - does this also put the new path in the navigation menu? I think it does. Any reason we can't leave it where it is, i.e. the old path, and then we won't move or break the existing navMenu entry? I prefer the settings to appear within the contribution configuration, it makes more sense to me there than in the generic settings with all the other system configs.

**Review by adixon — COMMENTED** [review]
Thanks for this, let me know what you think of my comments.

File: CRM/Iats/Utils.php
[the remark this answers is not in the document]
Comment (seamuslee001) [reply CRM/Iats/Utils.php:1]: @adixon I don't see a file here https://github.com/iATSPayments/com.iatspayments.civicrm/tree/master/CRM/Iats you might be getting confused with the class here https://github.com/iATSPayments/com.iatspayments.civicrm/blob/master/iats.civix.php#L9

File: iats.php [hunk 3]
```diff
@@ -195,7 +195,7 @@ function iats_civicrm_navigationMenu(&$navMenu) {
     'settings_page' => array(
       'label'      => 'iATS Payments Settings',
       'name'       => 'iATS Payments Settings',
-      'url'        => 'civicrm/admin/contribute/iatssettings',
+      'url'        => 'civicrm/admin/setting/iats',
```
[the remark this answers is not in the document]
Comment (seamuslee001) [reply iats.php:198]: @adixon no as the parent is still the same it just changes the url for the settings as far as I know

File: iats.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (seamuslee001) [reply iats.php:296]: @adixon this is a good question I hadn't thought about using the validate_callback functionality but perhaps that should be what is used here. I guess i was mostly focusing on the conversion from the custom quickform form to the generic form.

File: xml/Menu/iats.xml (hunk 2 above)
Comment (seamuslee001) [reply xml/Menu/iats.xml:16]: @adixon in regards to the url the Generic form assumes the last part of the url is the extension key https://github.com/civicrm/civicrm-core/blob/master/CRM/Admin/Form/SettingTrait.php#L142 I suppose that we could just do admin/contribute/iats but not sure if that woks and the format of the url is the same as in the documentation https://docs.civicrm.org/dev/en/latest/framework/setting/extension/

File: CRM/Iats/Utils.php
[the remark this answers is not in the document]
Comment (adixon) [reply CRM/Iats/Utils.php:1]: Nope, I definitely saw a Utils.php file and it was different. But it's not there now! I guess I was in a parallel universe. I do love a good Utils.php class! But I think I was seeing it in a checkout of your previous version of this PR, so it's all good ...

File: iats.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (adixon) [reply iats.php:296]: Okay, we'll call it an interim step, I'm happy to leave as is for now.

**Review by adixon — APPROVED** [review]

**Comment by adixon** [comment]
Thanks @seamuslee001

File: CRM/Iats/Utils.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    $hasNotbeenMigrated = CRM_Core_DAO::singleValueQuery("SELECT id FROM civicrm_setting WHERE name = 'iats_settings'");
+    if ($hasNotbeenMigrated) {
+      return Civi::settings()->get('iats_settings');
+    }
+    else {
+      foreach ($settings as $setting) {
+        $settingValue = Civi::settings()->get('iats_' . $setting);
+        if ($setting === 'days') {
+          // if setting value is blank set it to be disabled.
+          if (trim($settingValue, CRM_Core_DAO::VALUE_SEPARATOR) === '') {
```
Comment (totten) [inline CRM/Iats/Utils.php:17]: JFYI, @seamuslee001, circa March 22, we started seeing test-failures like this with the git-based test-runs of `com.iatspayment.civicrm` (across all tested versions of `civicrm-core` and `php`):

```
CRM_Iats_SettingsTest::testGetDates
trim() expects parameter 1 to be string, array given

/home/homer/buildkit/build/build-0/web/sites/all/modules/civicrm/ext/com.iatspayments.civicrm/CRM/Iats/Utils.php:17
/home/homer/buildkit/build/build-0/web/sites/all/modules/civicrm/ext/com.iatspayments.civicrm/tests/phpunit/CRM/Iats/SettingsTest.php:56
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2307
```

[link](https://test.civicrm.org/job/CiviCRM-Ext-Matrix/BKPROF=min,BUILDTYPE=git,CIVIVER=master,EXTKEY=com.iatspayments.civicrm,label=bknix-tmp/4830/testReport/(root)/CRM_Iats_SettingsTest/testGetDates/)

**Comment by KarinG** [comment]
@totten - thank you - I've actually put up a PR to remove the tests from current iATS Payments extension - as we have better coverage for iATS Payments within WFC testing now. Also there will be a new iATS Payments extension (ETA Fall/Winter) that will have its own Github Actions built in. 

So it's here - waiting on Alan to merge it - I don't self-merge :-)
https://github.com/iATSPayments/com.iatspayments.civicrm/pull/432

**Comment by adixon** [comment]
Actually, I believe that test is useful, it's not noise! "iats_days" setting is an array, not a string.

New issue:

https://github.com/iATSPayments/com.iatspayments.civicrm/issues/450

**Pull request merged**